src/lib.rs

```diff
@@ -1154,14 +1555,154 @@ where
+    }
+
+    fn propose_max_round_time(&self) -> Duration {
+        self.inner.config.propose_max_round_time
+    }
+
+    fn max_transactions(&self) -> NonZeroUsize {
+        self.inner.config.max_transactions
+    }
+
+    fn min_transactions(&self) -> usize {
+        self.inner.config.min_transactions
+    }
+
+    /// Generates and encodes a vote token
+
+    async fn should_start_round(&self, _: TYPES::Time) -> bool {
+        false
+    }
+
+    async fn send_event(&self, event: Event<TYPES, I::Leaf>) {
+        debug!(?event, "send_event");
+        let mut event_sender = self.inner.event_sender.write().await;
+        if let Some(sender) = &*event_sender {
+            if let Err(e) = sender.send_async(event).await {
+                error!(?e, "Could not send event to event_sender");
+                *event_sender = None;
+            }
+        }
+    }
+
+    fn public_key(&self) -> &TYPES::SignatureKey {
+        &self.inner.public_key
+    }
+
+    fn private_key(&self) -> &<TYPES::SignatureKey as SignatureKey>::PrivateKey {
+        &self.inner.private_key
+    }
+
+    async fn store_leaf(
+        &self,
+        old_anchor_view: TYPES::Time,
+        leaf: I::Leaf,
+    ) -> std::result::Result<(), hotshot_types::traits::storage::StorageError> {
+        let view_to_insert = StoredView::from(leaf);
+        let storage = &self.inner.storage;
+        storage.append_single_view(view_to_insert).await?;
+        storage.cleanup_storage_up_to_view(old_anchor_view).await?;
+        storage.commit().await?;
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl<
+        TYPES: NodeType<ConsensusType = ValidatingConsensus>,
+        I: NodeImplementation<
+            TYPES,
+            Leaf = ValidatingLeaf<TYPES>,
+            ConsensusMessage = ValidatingMessage<TYPES, I>,
+        >,
+    > hotshot_consensus::ValidatingConsensusApi<TYPES, I::Leaf, I>
+    for HotShotValidatingConsensusApi<TYPES, I>
+where
+    I::Exchanges: ValidatingExchangesType<TYPES, Message<TYPES, I>>,
+{
+    async fn send_direct_message<
+        PROPOSAL: ProposalType<NodeType = TYPES>,
+        VOTE: VoteType<TYPES>,
+    >(
+        &self,
+        recipient: TYPES::SignatureKey,
+        message: ValidatingMessage<TYPES, I>,
+    ) -> std::result::Result<(), NetworkError> {
+        let inner = self.inner.clone();
+        debug!(?message, ?recipient, "send_direct_message");
+        async_spawn_local(async move {
+            inner
+                .exchanges
+                .quorum_exchange()
+                .network()
+                .direct_message(
+                    Message {
+                        sender: inner.public_key.clone(),
+                        kind: MessageKind::from_consensus_message(message),
+                        _phantom: PhantomData,
+                    },
+                    recipient,
+                )
+                .await
+        });
+        Ok(())
+    }
+
+    // TODO remove and use exchange directly.
```

Comment:
Do we have an issue for this?

Reply:
Nope, and the description here isn't very accurate. I created an [issue](https://github.com/EspressoSystems/HotShot/issues/1194).